### ISSUE TITLE
RFC: Alternate signature for RouteHandler

### ISSUE THREAD

**Issue by gitblit** [decebals/pippo#91]
How do you feel about adopting one of the following signatures:

```java
void handleRoute(Context context, RouteHandlerChain chain);
```

or my preference

```java
void handleRoute(Context context);
```

where `Context` is an interface like:
```java
interface Context {
    Request request();
    Response response();
}
```

If we put the chain inside the `Context` then `Context` would need a `next()` method.

```java
void handleRoute(Context context) {
    if (someCondition) {
        context.next();
    }
}
```

I would drop the `Request` and `Response` factories and implement a `ContextFactory`.

I would also move most/all functionality out of `Request` and `Response` into `Context` and not wrap HttpServletRequest/Response at all.

```java
interface Context {
    HttpServletRequest request();
    HttpServletResponse response();
    void next();
    // refactored methods from Request
    // refactored methods from Response
}
```

We have alot of boilerplate with passing around three args all the time.  Customization often requires having both `Request` and `Response` so I would prefer to merge all these concepts into one type/argument.  From what I see most other frameworks have also come to this conclusion: Wicket, Ninja, Ratpack, etc.

I can live with two args if you are opposed to putting the chain inside the `Context`.

**Comment by decebals** [comment]
Personally I prefer to materialize both concepts `Request` and `Response` and to not hide the Request-Response nature. 
I don't want to be another Play, Ninja implementation and to clone these frameworks. 
Pippo is something different, you can read some parameters from the request and to send the response directly. In ninja for example, I feel that Context it's like a __DTO__, a big bucket, a storage for all response parameters, helper methods for both request and response. You cannot send a response directly (the framework creates a response from a Context and send this response, you cannot access the response) - . 

Sure we can simplify the signature of `RouteHandlerChain` if you think that are two many parameters with something like:
```java
public void handleRoute(RouteContext context) { ... }
```

but in this case in a regular handler I must write:
```java
public void handleRoute(RouteContext context) { 
    Request request = context.request();
    Response response = context.response();
    context.next();
}
```

but I don't see the value because we simplified the signature of method but we complicated the body. This `RouteContext` object maybe is valuable in the future if we want to put some business (methods) - of course is not ok to add more parameters in handleRoute and in this scenario we must create this context object (bucket).

**Comment by gitblit** [comment]
If you don't move methods out of `Response` and `Request` and into `RouteContext` then I agree, there is no value and it makes `RouteHandler` implementations more complex.

I also agree that `RouteContext` is a big bucket.  The current solution is three smaller buckets - and two of them are like ying & yang - intertwined/codependent.

I think your most important remark concerns response sending which would be unaffected by this reorganization.

I presently have two projects using Pippo.  One uses stock Pippo for a really light-weight app.  The other uses Pippo as a base for something similar to Play/Ninja.  So I'm not trying to make *Pippo* itself into Play/Ninja but I am trying to eliminate complexity and maximize code-reuse when marrying Pippo into something larger that is like Play/Ninja.

It's fine if you want to stick with the 3 args, I'll manage.  :)

**Comment by decebals** [comment]
I think that `Request`, `Response` concepts are very usefully in pippo because HTTP is a request-response protocol and as a user I wish to have a nice representation for these main concepts. I considerate that we done a nice job with the current implementations.
Also pippo is a micro web service that in essence is specialized in mixing server generated pages with RESTful APIs. 
Nowadays I see a real interest for this kind of library in java. On github and you can see that sparkjava has around 2K starts and ninjaframework/wicket only 650 (ninja has some extra years) because sparkjava it's simple, it's small, with few dependencies, it's for the mainstream people. Ninja/Play/Wicket are for engineers, has a big learning curve and contain a lot of dependencies because with these frameworks you can create big/complex applications. I believe that our niche is the first one because it's a real demand and an empty space. If I was happy with that offer these frameworks would not be there this pippo.
I am very happy that you have two applications developed with pippo because we have some feedbacks.  I would like to know what problems you have with the current approach. Also maybe I can't see the big picture, maybe you can give me more information about a possible implementation (some snippet code) and what value it brings. For now I see that two main concepts disappears (Request, Response)  and in their place comes a `RouteContext` with tens of methods and without a segmentation (request methods, response methods, ...). 
The solution with a RouteHandler in this form:
```java
public void handleRoute(RouteContext context) { 
    Request request = context.request();
    Response response = context.response();
    context.next();
}
```
it's ok for me if you think that the number of parameters in `RouteHandler.handleRoute()` is a problem (if we migrate to java 8 in the future I don't see any problems with the current signature of the method).

I appreciate very much our collaboration and I think that pippo is valuable and we do a great work.

**Comment by gitblit** [comment]
Closed by #95.
